vmap support for torch.tril and torch.triu

isdanni opened the pull request:
Summary:
Add vmap support for torch.tril and torch.triu.

Fix: #91403

Test Plan: GitHub pipeline

Differential Revision: D43016624

### Expected behavior
Same as using for-loop:

```python
import torch 

x = torch.randn(32, 3)
results = []
for xi in x:
  y = torch.triu(xi)
  results.append(y)
"""
triu: input tensor must have at least 2 dimensions
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-7-d726203efb0e> in <module>
      4 results = []
      5 for xi in x:
----> 6   y = torch.triu(xi)
      7   results.append(y)
RuntimeError: triu: input tensor must have at least 2 dimensions
"""
```

Skylion007 commented on aten/src/ATen/functorch/BatchRulesViews.cpp:
```diff
@@ -556,12 +556,28 @@ Tensor trace_decomp(const Tensor& tensor) {
+    int64_t diagonal = 0) {
+  TORCH_CHECK(self.dim() >= 2, "tril: The input tensor must have at least 2 dimensions.");
+  auto result = at::tril(self, diagonal);
+  return std::make_tuple(result, 0);
```
```suggestion
  return std::make_tuple(std::move(result), 0);
```

lezcano replied:
This optimisation is almost always negligible in the grand scheme of things. If we really want to enforce it all across we should add it to the linter.

Skylion007 replied:
@lezcano Adding it to a linter is bit more difficult that it sounds, a lot of false positives / negatives. Further complicated by template forwarding etc.

lezcano replied:
Then I think it's not worth it to reviewing PRs making this point. As discussed, all these optimisations are not useful and not worth the CI run unless they are in a very hot path.

Skylion007 replied:
@lezcano there is work on a clang-tidy check that I've been experimenting with, but it's not quite ready yet: https://reviews.llvm.org/D137205

Skylion007 commented on aten/src/ATen/functorch/BatchRulesViews.cpp:
```diff
@@ -556,12 +556,28 @@ Tensor trace_decomp(const Tensor& tensor) {
+    int64_t diagonal = 0) {
+  TORCH_CHECK(self.dim() >= 2, "triu: The input tensor must have at least 2 dimensions.");
+  auto result = at::triu(self, diagonal);
+  return std::make_tuple(result, 0);
```
```suggestion
  return std::make_tuple(std::move(result), 0);
```

Skylion007 reviewed (commented):
Copying both these tensors includes an atomic lock and ref count inc / dec that can be avoided with an std::move().

zou3519 commented:
@pytorchbot rebase

isdanni committed:
vmap support for torch.tril and torch.triu (#94287)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/94287

Add vmap support for torch.tril and torch.triu.

Issue: https://github.com/pytorch/pytorch/issues/91403

Test Plan: GitHub pipeline

Reviewed By: zou3519

Differential Revision: D43016624

fbshipit-source-id: 6e73ce6f1c83be8bafc70a039c06404b60180a87
